chain/syncer_status.go

```diff
@@ -0,0 +1,173 @@
+	synTip types.TipSetKey, synHeight uint64, synTrust bool,
+	synFetchedToHeight uint64, synFetchComplete bool, synStart time.Time) SyncerStatus {
+	return SyncerStatus{
+		ValidatedHead:          valHead,
+		ValidatedHeadHeight:    valHeight,
+		SyncingTip:             synTip,
+		SyncingHeight:          synHeight,
+		SyncingTrusted:         synTrust,
+		SyncingFetchedToHeight: synFetchedToHeight,
+		SyncingFetchComplete:   synFetchComplete,
+		SyncStarted:            synStart,
+	}
+}
+
+// SyncStatusReport when started will record:
+// - current tipsetkey and height of the heaviest synced chain
+// - time it was started
+// - new tipsetkey and heigh of chain being synced
+// - if its perfroming a trusted sync
+// - if its completed syncing
+// - poll the fetcher for the hieght of the tipset currently being fetched
+type SyncStatusReporter struct {
+	chainReader syncerStatusChainReader
+	clock       clock.Clock
+	fetchNotifi syncerStatusFetchNotifi
+
+	recording bool
+	done      chan struct{}
+
+	statusMu sync.Mutex
+	status   SyncerStatus
+}
+
+// NewSyncStatusReporter returns a new SyncStatusReporter.
+func NewSyncStatusReporter(cr syncerStatusChainReader, fn syncerStatusFetchNotifi, c clock.Clock) *SyncStatusReporter {
+	valHead, err := cr.GetTipSet(cr.GetHead())
+	if err != nil {
+		panic(err)
+	}
+	valHeight, err := valHead.Height()
+	if err != nil {
+		panic(err)
+	}
+	return &SyncStatusReporter{
+		chainReader: cr,
+		clock:       c,
+		fetchNotifi: fn,
+
+		done:      make(chan struct{}),
+		recording: false,
+
+		status: SyncerStatus{
+			ValidatedHead:          valHead.Key(),
+			ValidatedHeadHeight:    valHeight,
+			SyncingTip:             types.UndefTipSet.Key(),
+			SyncingHeight:          0,
+			SyncingTrusted:         false,
+			SyncingFetchedToHeight: 0,
+			SyncingFetchComplete:   true,
+			SyncStarted:            c.Now(),
+		},
+	}
+}
+
+// Start sets the status values and starts a go routine subscribed to fetcher notifications for height changes.
+func (s *SyncStatusReporter) Start(synTip types.TipSetKey, synHeight uint64, synTrust bool) error {
+	if s.recording {
+		panic("cannot call start again on sync status reporter developer error")
+	}
+	s.recording = true
+
+	valHead, err := s.chainReader.GetTipSet(s.chainReader.GetHead())
+	if err != nil {
+		return err
+	}
+	valHeight, err := valHead.Height()
+	if err != nil {
+		return err
+	}
+	s.statusMu.Lock()
+	s.status = SyncerStatus{
+		ValidatedHead:          valHead.Key(),
+		ValidatedHeadHeight:    valHeight,
+		SyncingTip:             synTip,
+		SyncingHeight:          synHeight,
+		SyncingTrusted:         synTrust,
+		SyncingFetchedToHeight: valHeight,
+		SyncingFetchComplete:   false,
+		SyncStarted:            s.clock.Now(),
+	}
+	s.statusMu.Unlock()
+	go func() {
+		fetchSub := s.fetchNotifi.Notifications().Sub(net.FetchTsTopic)
+		defer s.fetchNotifi.Notifications().Unsub(fetchSub, net.FetchTsTopic)
+		for {
+			select {
+			case raw := <-fetchSub:
```

Comment:
An alternative to this could be a ticker that queries a method on the fetcher for its current highest fetched tipset.